File: exclusive-time-of-functions/shaun-lt.cpp
//runtime 56ms, faster than 17%, mem less than 10%
class Solution {
public:
    vector<int> exclusiveTime(int n, vector<string>& logs) {
        stack<int> st;
        vector<int> res(n);
        if (logs.empty()) return res;
        vector<string> s = split(logs[0], ':');
        int id = stoi(s[0]), ts = stoi(s[2]);
        st.push(id);
        int prev = ts;
        for (int i = 1; i < logs.size(); i++) {
            s = split(logs[i], ':');
            id = stoi(s[0]);
            ts = stoi(s[2]);
            if (s[1] == "start") {
                if (!st.empty())
                    res[st.top()] += ts - prev;
                st.push(id);
                prev = ts;
            }else {
                res[st.top()] += ts - prev + 1;
                st.pop();
                prev = ts +1;
                
            }
        }
        return res;

    }
private:
    vector<string> split (const string &s, char delim) {
        vector<string> result;
        stringstream ss (s);

        string token;
        while (getline (ss, token, delim)) { //getline delim only supports char type
            result.push_back (token);
        }

        return result;
    }

};

File: exclusive-time-of-functions/shaun-lt.java
//runtime 12ms, faster than 93%, mem less than 7%
//time O(n), space O(1)
public class Solution {
    public int[] exclusiveTime(int n, List < String > logs) {
        
        Stack<Integer> stack = new Stack<>();
        int [] res = new int[n];
        
        //function_id: start|end: timestamp
        String [] s = logs.get(0).split(":");
        int id = Integer.parseInt(s[0]), ts = Integer.parseInt(s[2]);
        stack.push(id); 
        int prev = ts;
        for (int i = 1; i < logs.size(); i++) {
            s = logs.get(i).split(":");
            id = Integer.parseInt(s[0]);
            ts = Integer.parseInt(s[2]);
            if (s[1].equals("start")) {
                if (!stack.isEmpty())
                    res[stack.peek()] += ts - prev;
                stack.push(id);
                prev = ts;
            }else {
                res[stack.peek()] += ts - prev +1;
                stack.pop();
                prev = ts + 1;
            }
        }
        return res;
    }
            
}
